### DISCUSSION TITLE
Unable to run this query through prisma

### DISCUSSION THREAD

**Discussion by nisha-singh-tredence** [prisma/prisma#22084] [Q&A] [answered]
### Question

Query is,
```sql
INSERT INTO tablename(emp_id, year, cost, "CREATED_DATE") 
VALUES 
(ENCODE(PGP_SYM_ENCRYPT(111::text, 'fghjkliytttttttsdcsfdsfdsf'), 'base64'),2024,'baf13e930109b39640e1c8dcb8c259f8', 'Thu Nov 23 2023 10:41:23 GMT+0530 (India Standard Time)'),(ENCODE(PGP_SYM_ENCRYPT(222::text, 'fghjkliytttttttsdcsfdsfdsf'), 'base64'),2024,'baf13e930109b39640e1c8dcb8c259f8', 'Thu Nov 23 2023 10:41:23 GMT+0530 (India Standard Time)');
```
Successfully Run in PG but it's failed when try to run through prisma.

Please guide how can run this query through Prisma
I have used Prisma Version:
   "prisma": "^4.5.0",

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @nisha-singh-tredence 👋 

Thank you for raising this question.

We do have a feature request to allow custom SQL to read or write a column's value.

- https://github.com/prisma/prisma/issues/9064

Also, as suggested in this previous [discussion](https://github.com/prisma/prisma/discussions/7798#discussioncomment-901008), you will need to use [raw query](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access) to achieve this.

For example, Let's say we have defined our Prisma schema to look like
```prisma
generator client {
  provider        = "prisma-client-js"
  binaryTargets   = ["native"]
  previewFeatures = ["postgresqlExtensions"]
}

datasource db {
  provider   = "postgresql"
  url        = env("DATABASE_URL")
  extensions = [pgcrypto]
}

model Test {
  emp_Id       String   @id
  year         Int
  cost         String
  CREATED_DATE DateTime
}
```
Notice the use of the `previewFeatures = ["postgresqlExtensions"]` and `extensions=[pgcrypto]`. This is to ensure we have access to the native database function like `PGP_SYM_ENCRYPT` and `ENCODE`. See this [guide](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/enable-native-database-functions) on enabling native database functions.

You can then run either `npx prisma migrate dev` or `npx prisma db push` to commit your changes.

Within your script, you can write this query
```js
const result = await prisma.$executeRaw`
INSERT INTO "Test" ("emp_Id", year, cost, "CREATED_DATE")
VALUES 
  (ENCODE(PGP_SYM_ENCRYPT('111'::text, 'fghjkliytttttttsdcsfdsfdsf'), 'base64'), 2024, 'baf13e930109b39640e1c8dcb8c259f8', '2023-11-23T10:41:23.000Z'),
  (ENCODE(PGP_SYM_ENCRYPT('222'::text, 'fghjkliytttttttsdcsfdsfdsf'), 'base64'), 2024, 'baf13e930109b39640e1c8dcb8c259f8', '2023-11-23T10:41:23.000Z');
`;
```
After running it, you should get this output below. **N.B** I have logging enabled. The `2` is just the number of rows affected. In this case, 2 records were inserted.
```
prisma:info Starting a postgresql pool with 17 connections.
prisma:query 
INSERT INTO "Test" ("emp_Id", year, cost, "CREATED_DATE")
VALUES 
  (ENCODE(PGP_SYM_ENCRYPT('111'::text, 'fghjkliytttttttsdcsfdsfdsf'), 'base64'), 2024, 'baf13e930109b39640e1c8dcb8c259f8', '2023-11-23T10:41:23.000Z'),
  (ENCODE(PGP_SYM_ENCRYPT('222'::text, 'fghjkliytttttttsdcsfdsfdsf'), 'base64'), 2024, 'baf13e930109b39640e1c8dcb8c259f8', '2023-11-23T10:41:23.000Z');

2
```
From here, I run a `findMany()` query to see the output
```
prisma:info Starting a postgresql pool with 17 connections.
prisma:query SELECT "public"."Test"."emp_Id", "public"."Test"."year", "public"."Test"."cost", "public"."Test"."CREATED_DATE" FROM "public"."Test" WHERE 1=1 OFFSET $1
[
  {
    emp_Id: 'ww0EBwMCJFVn6lhEQXVx0jQBLjQPrytAIHDZ+OrjmFHseXygw1kx5iC60tSu6+Pl8AT9eZasow84\n' +
      'y5zVazOVPR5flE2B',
    year: 2024,
    cost: 'baf13e930109b39640e1c8dcb8c259f8',
    CREATED_DATE: 2023-11-23T10:41:23.000Z
  },
  {
    emp_Id: 'ww0EBwMCGKesoCJ01r1+0jQBOvUCarsd/J/nhRyMfWMeWt5jz2jhaBZ2tibeBo74WxyFfAGdQtiT\n' +
      '3cD24bIc3JVJfp+U',
    year: 2024,
    cost: 'baf13e930109b39640e1c8dcb8c259f8',
    CREATED_DATE: 2023-11-23T10:41:23.000Z
  }
]
```
If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by nisha-singh-tredence** [comment, marked as answer]
Hi @ludralph 

Thanks for reply,

I have tried same to run
```js
const rawInsertQuery = INSERT INTO tableName ("SF_EMP_ID", "YEAR", "COST", "CREATED_DATE") 
    VALUES 
        (ENCODE(PGP_SYM_ENCRYPT('121', 'YTRFCDTDH678FFFTGGGHHH''),'base64'),2024,'baf13e930109b39640e1c8dcb8c259f8', '2023-11-23T06:54:28.780Z'),
        (ENCODE(PGP_SYM_ENCRYPT('4169', 'YTRFCDTDH678FFFTGGGHHH'),'base64'),2024,'baf13e930109b39640e1c8dcb8c259f8', '2023-11-23T06:54:28.780Z');
```
IN PRISMA Schema File,
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["fullTextSearch", "postgresqlExtensions"]
  binaryTargets   = ["native", "debian-openssl-1.1.x"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  extensions = [pgcrypto]
}
```
successfully generated and db pushed ,

But when try to execute 
await prisma.$executeRaw`${rawInsertQuery}`

It's throw error

{
    "data": {
        "code": "P2010",
        "clientVersion": "4.5.0",
        "meta": {
            "code": "42601",
            "message": "db error: ERROR: syntax error at or near \"$1\""
        }
    }
}

Please have a look

**Reply by ludralph** [reply]
Hi @nisha-singh-tredence 

You had a typo in your query near `
        (ENCODE(PGP_SYM_ENCRYPT('121', 'YTRFCDTDH678FFFTGGGHHH'')`. Notice the additional `'` before the closing`)`.


You can try this query 

```js
const rawInsertQuery = Prisma.sql`INSERT INTO tableName ("SF_EMP_ID", "YEAR", "COST", "CREATED_DATE") 
  VALUES 
    (ENCODE(PGP_SYM_ENCRYPT('121', 'YTRFCDTDH678FFFTGGGHHH'), 'base64'),2024,'baf13e930109b39640e1c8dcb8c259f8', '2023-11-23T06:54:28.780Z'),
    (ENCODE(PGP_SYM_ENCRYPT('4169', 'YTRFCDTDH678FFFTGGGHHH'), 'base64'),2024,'baf13e930109b39640e1c8dcb8c259f8', '2023-11-23T06:54:28.780Z');`

    const res = await prisma.$executeRaw`${rawInsertQuery}`
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by nisha-singh-tredence** [comment]
Thanks!
